File: 973. K Closest Points to Origin/k_closest.rb
# Runtime: 384 ms, faster than 34.69% of Ruby online submissions for K Closest Points to Origin.
# Memory Usage: 214.3 MB, less than 40.82% of Ruby online submissions for K Closest Points to Origin
# @param {Integer[][]} points
# @param {Integer} k
# @return {Integer[][]}
def k_closest(points, k)
  ans = []
  for p in points
      square_dist = p[0]*p[0] + p[1]*p[1]
      ans << [square_dist, p]
  end
  ans.sort!
  ans[..k-1].map{|x| x[1]}
end


# Runtime: 236 ms, faster than 93.88% of Ruby online submissions for K Closest Points to Origin.
# Memory Usage: 211.9 MB, less than 71.43% of Ruby online submissions for K Closest Points to Origin.
def k_closest(points, k)
  points.sort_by!{|p| p[0]*p[0] + p[1]*p[1]}[..k-1]
end

# Runtime: 232 ms, faster than 95.92% of Ruby online submissions for K Closest Points to Origin.
# Memory Usage: 211.9 MB, less than 69.39% of Ruby online submissions for K Closest Points to Origin.
def k_closest(points, k)
  points.sort_by!{|p| p[0]*p[0] + p[1]*p[1]}.first(k)
end

File: 973. K Closest Points to Origin/k_closest.kt
// Runtime: 755 ms, faster than 94.91% of Kotlin online submissions for K Closest Points to Origin.
// Memory Usage: 60.7 MB, less than 97.22% of Kotlin online submissions for K Closest Points to Origin.
fun kClosest(points: Array<IntArray>, k: Int): Array<IntArray> {
    val n = points.size
    val distances: Array<IntArray> = Array(n) {intArrayOf(0,0)}
    for (i in 0 until n) {
        var x = points[i][0]
        var y = points[i][1]
        var distance = x*x + y*y
        distances[i] = intArrayOf(distance, i)
    }

    distances.sortWith(compareBy({ it[0]}, { it[1]}))

    val result: Array<IntArray> = Array(k) {intArrayOf(0,0)}
    for (i in 0 until k) {
        result[i] = points[distances[i][1]]
    }

    return result
}


fun kClosest2(points: Array<IntArray>, k: Int): Array<IntArray> =
    points.sortedBy{it[0]*it[0]+it[1]*it[1]}.take(k).toTypedArray()


fun main() {

}